iroh-resolver/src/content_loader.rs

```diff
@@ -0,0 +1,258 @@
+    pub indexer: Option<Url>,
+    pub http_gateways: Vec<GatewayUrl>,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum GatewayUrl {
+    Full(Url),
+    Subdomain(String),
+}
+
+impl FromStr for GatewayUrl {
+    type Err = anyhow::Error;
+    fn from_str(input: &str) -> Result<Self> {
+        if input.starts_with("http") || input.starts_with("https") {
+            let url = input.parse()?;
+            return Ok(GatewayUrl::Full(url));
+        }
+
+        Ok(GatewayUrl::Subdomain(input.to_string()))
+    }
+}
+
+impl GatewayUrl {
+    pub fn as_string(&self) -> String {
+        match self {
+            GatewayUrl::Full(url) => url.to_string(),
+            GatewayUrl::Subdomain(s) => s.clone(),
+        }
+    }
+
+    pub fn as_url(&self, cid: &Cid) -> Url {
+        let cid_str = cid.to_string_of_base(Base::Base32Lower).unwrap();
+        match self {
+            GatewayUrl::Full(raw) => {
+                let mut url = raw.join(&cid_str).unwrap();
+                url.set_query(Some("format=raw"));
+                url
+            }
+            GatewayUrl::Subdomain(raw) => format!("https://{}.ipfs.{}?format=raw", cid_str, raw)
+                .parse()
+                .unwrap(),
+        }
+    }
+}
+
+impl FullLoader {
+    pub fn new(client: Client, config: FullLoaderConfig) -> Result<Self> {
+        let indexer = config.indexer.map(Indexer::new).transpose()?;
+
+        Ok(Self {
+            client,
+            indexer,
+            http_gateways: config.http_gateways,
+        })
+    }
+
+    /// Fetch the next gateway url, if configured.
+    async fn next_gateway(&self) -> Option<&GatewayUrl> {
+        // TODO: maybe roundrobin?
+        if self.http_gateways.is_empty() {
+            return None;
+        }
+        let gw = self.http_gateways.choose(&mut rand::thread_rng()).unwrap();
+        Some(gw)
+    }
+
+    async fn fetch_store(&self, cid: &Cid) -> Result<Option<LoadedCid>> {
+        match self.client.try_store() {
+            Ok(store) => Ok(store.get(*cid).await?.map(|data| LoadedCid {
+                data,
+                source: Source::Store(IROH_STORE),
+            })),
+            Err(err) => {
+                info!("No store available: {:?}", err);
+                Ok(None)
+            }
+        }
+    }
+
+    async fn fetch_bitswap(&self, ctx: ContextId, cid: &Cid) -> Result<Option<LoadedCid>> {
+        match self.client.try_p2p() {
+            Ok(p2p) => {
+                let providers: HashSet<_> = if let Some(ref indexer) = self.indexer {
+                    if let Ok(providers) = indexer.find_providers(*cid).await {
+                        providers.into_iter().map(|p| p.id).collect()
+                    } else {
+                        Default::default()
+                    }
+                } else {
+                    Default::default()
+                };
+
+                let data = p2p.fetch_bitswap(ctx.into(), *cid, providers).await?;
```

Comment:
`find_providers` here is returning all providers - you can filter on `Metadata` to find only the providers that are saying they work with Bitswap - if you don't filter you'll also get fil-markets providers that only speak graphsync

the `Metadata` is base64 encoding of a uvarint encoding of (0x900) - https://github.com/multiformats/multicodec/blob/master/table.csv#L138

Reply:
fixed :)